src/Tes.ApiClients/HttpApiClient.cs

```diff
@@ -68,6 +69,36 @@ protected HttpApiClient(CachingRetryHandler cachingRetryHandler, ILogger logger)
         /// </summary>
         protected HttpApiClient() { }
 
+        /// <summary>
+        /// A logging Polly retry handler.
+        /// </summary>
+        /// <param name="caller">Calling method name.</param>
+        /// <returns><see cref="RetryHandler.OnRetryHandler"/></returns>
+        private RetryHandler.OnRetryHandler LogRetryErrorOnRetryHandler([System.Runtime.CompilerServices.CallerMemberName] string? caller = default)
+            => new((exception, timeSpan, retryCount, correlationId) =>
+            {
+                Logger?.LogError(exception, @"Retrying in {Method}: RetryCount: {RetryCount} TimeSpan: {TimeSpan} CorrelationId: {CorrelationId:D}", caller, retryCount, timeSpan, correlationId);
+            });
+
+        /// <summary>
+        /// A logging Polly retry handler.
+        /// </summary>
+        /// <typeparam name="TResult">See <see cref="PolicyBuilder{TResult}"/></typeparam>
+        /// <param name="caller">Calling method name.</param>
+        /// <returns><see cref="RetryHandler.OnRetryHandler{TResult}"/></returns>
+        private RetryHandler.OnRetryHandler<TResult> LogRetryErrorOnRetryHandler<TResult>([System.Runtime.CompilerServices.CallerMemberName] string? caller = default)
+            => new((result, timeSpan, retryCount, correlationId) =>
+            {
+                if (result.Exception is null)
```

Comment:
What are the cases where the exception is null?  I find it kind of weird that Polly would retry without an exception.

Reply:
The Polly policy builders that take a generic type argument (like the one returned by `HttpPolicyExtensions.HandleTransientHttpError()`) retriable failures can be defined by state where no exceptions are involved (including conditions added by `OrResult()`). In those cases, exception will be `null` in the retry handler.